subsys/bluetooth/services/cgms/cgms.c

```diff
@@ -0,0 +1,537 @@
+				BT_GATT_PERM_READ_AUTHEN | BT_GATT_PERM_WRITE_AUTHEN),
+	/* CGM Specific Ops Control Point */
+	BT_GATT_CHARACTERISTIC(BT_UUID_CGM_SPECIFIC_OPS_CONTROL_POINT,
+				BT_GATT_CHRC_WRITE | BT_GATT_CHRC_INDICATE,
+				BT_GATT_PERM_WRITE_AUTHEN,
+				NULL, socp_on_receive, NULL),
+	/* CGM Specific Ops Control Point CCCD */
+	BT_GATT_CCC(cgms_socp_ccc_cfg_changed,
+				BT_GATT_PERM_READ_AUTHEN | BT_GATT_PERM_WRITE_AUTHEN),
+);
+
+static void bt_cgms_notify_meas(struct bt_conn *conn, void *data)
+{
+	struct cgms_meas *meas = (struct cgms_meas *)data;
+	uint8_t meas_size;
+	struct net_buf_simple *meas_buf = NET_BUF_SIMPLE(CGMS_MEAS_LENGTH);
+
+	net_buf_simple_init(meas_buf, sizeof(meas_size));
+
+	net_buf_simple_add_u8(meas_buf, meas->flag);
+	net_buf_simple_add_le16(meas_buf, meas->glucose_concentration);
+	net_buf_simple_add_le16(meas_buf, meas->time_offset);
+	if (meas->sensor_status_annunciation.status != 0) {
+		net_buf_simple_add_u8(meas_buf, meas->sensor_status_annunciation.status);
+	}
+	if (meas->sensor_status_annunciation.calib_temp != 0) {
+		net_buf_simple_add_u8(meas_buf, meas->sensor_status_annunciation.calib_temp);
+	}
+	if (meas->sensor_status_annunciation.warning != 0) {
+		net_buf_simple_add_u8(meas_buf, meas->sensor_status_annunciation.warning);
+	}
+
+	meas_size = meas_buf->len + 1;
+	net_buf_simple_push_u8(meas_buf, meas_size);
+
+	/* If conn is NULL, it implies this is a periodic notification.
+	 * Send it to all peers.
+	 * If conn is set, verify if the notification of this conn is enabled,
+	 * and send notification accordingly.
+	 */
+	if (conn == NULL) {
+		bt_gatt_notify(NULL, &cgms_svc.attrs[CGMS_SVC_MEAS_ATTR_IDX],
+			meas_buf->data, meas_size);
+	} else if (bt_gatt_is_subscribed(conn, &cgms_svc.attrs[CGMS_SVC_MEAS_ATTR_IDX],
+			BT_GATT_CCC_NOTIFY)) {
+		bt_gatt_notify(conn, &cgms_svc.attrs[CGMS_SVC_MEAS_ATTR_IDX],
+			meas_buf->data, meas_size);
+	} else {
+		LOG_INF("Client disabled the measurement notification");
+	}
+}
+
+int cgms_racp_send_response(struct bt_conn *peer, struct net_buf_simple *rsp)
+{
+	static struct bt_gatt_indicate_params indicate_data;
+
+	indicate_data.attr = &cgms_svc.attrs[CGMS_SVC_RACP_ATTR_IDX];
+	indicate_data.data = rsp->data;
+	indicate_data.len = rsp->len;
+
+	return bt_gatt_indicate(peer, &indicate_data);
+}
+
+int cgms_racp_send_record(struct bt_conn *peer, struct cgms_meas *entry)
+{
+	bt_cgms_notify_meas(peer, entry);
+	return 0;
+}
+
+int cgms_socp_send_response(struct bt_conn *peer, struct net_buf_simple *rsp)
+{
+	static struct bt_gatt_indicate_params indicate_data;
+
+	indicate_data.attr = &cgms_svc.attrs[CGMS_SVC_SOCP_ATTR_IDX];
+	indicate_data.data = rsp->data;
+	indicate_data.len = rsp->len;
+
+	return bt_gatt_indicate(peer, &indicate_data);
+}
+
+int bt_cgms_measurement_add(struct bt_cgms_meas measurement)
+{
+	/*Prepare the measurement */
+	struct cgms_meas meas;
+
+	if (cgms_inst.status.status.status & CGMS_MEAS_STATUS_SESSION_STOPPED) {
```

Comment:
It seems to me that `status` can be used in different contexts. So it should be of type `atomic_t`.